### PR TITLE
Add TypeScript definition

### LINKED ISSUE #74: Add TypeScript type definition
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/pify/issues/74)

Would be nice to have TypeScript type definitions bundled in this package. The types should be strict and should be generic so it preserves the original function type.

The type definition should follow [this styleguide](https://github.com/sindresorhus/typescript-definition-style-guide), and should have full docs and type tests.

Some examples:
- https://github.com/sindresorhus/semver-regex/commit/039944bb5b3876913171b895488874b5c200e458
- https://github.com/sindresorhus/global-dirs/commit/0aae4d64887f77af94c4fb7ca12aa9f93c0a8291
- https://github.com/sindresorhus/md5-hex/commit/bbbfeff840f26973fbffc16d4863ae6d120079f3

## Note that this is not an easy task and requires you to have advanced TypeScript experience. There have been multiple attempts and a lot of wasted time on my part in the past. If you intend to work on this, please look at the previous attempts and the feedback given there.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $100.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($80.00)
- [omgimalexis omgimalexis](https://issuehunt.io/u/omgimalexis) ($20.00)

### Submitted pull Requests
- [#75 [WIP] Add TypeScript definition](https://issuehunt.io/r/sindresorhus/pify/pull/75)
- [#76 Add index.d.ts and tests](https://issuehunt.io/r/sindresorhus/pify/pull/76)
- [#85 Add TypeScript definition](https://issuehunt.io/r/sindresorhus/pify/pull/85)
- [#86 Add TypeScript definitions](https://issuehunt.io/r/sindresorhus/pify/pull/86)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/pify/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by satyarohith**
closes #74

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#74: Add TypeScript type definition](https://issuehunt.io/repos/41585934/issues/74)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by satyarohith** [comment]
@sindresorhus Would love your feedback. :)

File: index.test-d.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+obj.method = (value: string): string => value;
+
+expectType<Promise<number>>(pify(add)(1, 2).then((result: number) => result));
+expectType<Function>(pify(add)(1, 2).then);
+expectType<(a: number, b: number) => Promise<number>>(pify(add));
+expectType<Function>(pify(obj));
+expectType<(value: string) => Promise<string>>(pify(obj.method));
+
+expectType<any>(pify(fs).readFile('package.json', 'utf8').then);
```
Comment (sindresorhus) [inline index.test-d.ts]: There needs to be a lot more tests. Especially to ensure the input function is exactly the same as the output other than it returning a promise.

See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/pify/pify-tests.ts for inspiration,

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
+interface PromiseModule {
+	// This piece of code is obtained from https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/pify/index.d.ts.
```
Comment (sindresorhus) [inline index.d.ts]: Instead of linking to where you got it from, the comment should explain what it does.

**Comment by sindresorhus** [comment]
It's a good start, but I think you should put more effort into it.

**Comment by sindresorhus** [comment]
Note from the issue:

> The type definition should follow [this styleguide](https://github.com/sindresorhus/typescript-definition-style-guide), and should have full docs and type tests.

> Some examples:
- https://github.com/sindresorhus/semver-regex/commit/039944bb5b3876913171b895488874b5c200e458
- https://github.com/sindresorhus/global-dirs/commit/0aae4d64887f77af94c4fb7ca12aa9f93c0a8291
- https://github.com/sindresorhus/md5-hex/commit/bbbfeff840f26973fbffc16d4863ae6d120079f3

**Comment by sindresorhus** [comment]
Use `unknown` whenever possible instead of `any`.

**Comment by sindresorhus** [comment]
Also note the docs changes in https://github.com/sindresorhus/pify/commit/e0a8b71e68e9101beaf272c1b77c664d1c09fc98

**Comment by sindresorhus** [comment]
@satyarohith Bump

**Comment by satyarohith** [comment]
Thanks for your patience! Just busy with some life events and university
exams. I'll take this soon after I finish them.  I don't mind if anyone
wants to work on top of this for now. :)

**Comment by chinesedfan** [comment]
@satyarohith If possible, I'd like to have a try. Thanks.

**Comment by satyarohith** [comment]
Closing this in favour of #76

**Pull request closed without merging**